### PR TITLE
Add config squiggle for intelliSenseMode/compilerPath mismatch

### PR DESCRIPTION
**Pull request by Colengms**

### CODE REVIEW

File: Extension/src/LanguageServer/configurations.ts [hunk 1]
```diff
@@ -751,6 +751,27 @@ export class CppProperties {
...
+            // Check if intelliSenseMode and compilerPath are compatible
+            if (os.platform() === 'win32') {
+                // cl.exe is only available on Windows
+                const intelliSenseModeStart: number = curText.search(/\s*\"intelliSenseMode\"\s*:\s*\"/);
+                if (intelliSenseModeStart !== -1) {
+                    const intelliSenseModeValueStart: number = curText.indexOf('"', curText.indexOf(":", intelliSenseModeStart));
+                    const intelliSenseModeValueEnd: number = intelliSenseModeStart === -1 ? -1 : curText.indexOf('"', intelliSenseModeValueStart + 1) + 1;
+
+                    // cl.exe and msvc mode should be used together
+                    let compilerPathAndArgs: util.CompilerPathAndArgs = util.extractCompilerPathAndArgs(this.CurrentConfiguration.compilerPath);
```
Comment (michelleangela) [inline Extension/src/LanguageServer/configurations.ts]: Can this check be factored out into a function that does something like:

`isCompilerIntelliSenseModeCompatible(osPlatform: string): boolean { ... }`

We would need to reuse the code that validates inputs for both UI and JSON later on.

File: Extension/src/LanguageServer/configurations.ts [hunk 2]
```diff
@@ -751,6 +751,27 @@ export class CppProperties {
...
+                    if (compilerPathAndArgs.compilerPath.endsWith("cl.exe") !== (this.CurrentConfiguration.intelliSenseMode === "msvc-x64")) {
+                        let message: string = `IntelliSenseMode ${this.CurrentConfiguration.intelliSenseMode} is incompatible with compilerPath.`;
+                        let diagnostic: vscode.Diagnostic = new vscode.Diagnostic(
+                            new vscode.Range(document.positionAt(curTextStartOffset + intelliSenseModeValueStart),
+                                document.positionAt(curTextStartOffset + intelliSenseModeValueEnd)),
+                            message, vscode.DiagnosticSeverity.Warning);
+                        diagnostics.push(diagnostic);
```
Comment (bobbrow) [inline Extension/src/LanguageServer/configurations.ts:847]: we should log telemetry for this case

**Review by bobbrow — CHANGES_REQUESTED** [review]

File: Extension/src/LanguageServer/configurations.ts (hunk 1 above)
Comment (Colengms) [reply Extension/src/LanguageServer/configurations.ts]: Added

File: Extension/src/LanguageServer/configurations.ts (hunk 2 above)
Comment (Colengms) [reply Extension/src/LanguageServer/configurations.ts:847]: Added

**Review by sean-mcmanus — APPROVED** [review]

File: Extension/src/LanguageServer/configurations.ts [hunk 3]
```diff
@@ -846,10 +875,8 @@ export class CppProperties {
             const compilerPathEnd: number = compilerPathStart === -1 ? -1 : curText.indexOf('"', curText.indexOf('"', curText.indexOf(":", compilerPathStart)) + 1) + 1;
 
             if (this.prevSquiggleMetrics[this.CurrentConfiguration.name] === undefined) {
-                this.prevSquiggleMetrics[this.CurrentConfiguration.name] = { PathNonExistent: 0, PathNotAFile: 0, PathNotADirectory: 0, CompilerPathMissingQuotes: 0 };
+                this.prevSquiggleMetrics[this.CurrentConfiguration.name] = { PathNonExistent: 0, PathNotAFile: 0, PathNotADirectory: 0, CompilerPathMissingQuotes: 0, CompilerModeMismatch: 0 };
```
Comment (bobbrow) [inline Extension/src/LanguageServer/configurations.ts]: let's move this up with the other variables too.

**Review by bobbrow — APPROVED** [review]

File: Extension/src/LanguageServer/configurations.ts (hunk 3 above)
Comment (Colengms) [reply Extension/src/LanguageServer/configurations.ts]: Moved

**Review by michelleangela — APPROVED** [review]

**Pull request merged**